### PR TITLE
(PUP-6383) Failover report acceptance

### PR DESCRIPTION
**Pull request by Magisus**
Still trying to get a build through Jenkins, but these did all pass with a slightly outdated build (from right before Branan's work got merged).

In the meantime, I'll just put these up for review.

### CODE REVIEW

**Comment by Magisus** [comment]
Update: these tests all pass with the latest version of the failover and failover report code.

File: acceptance/tests/reports/failover_master.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+test_name "The report specifies which master was contacted during failover" do
+  master_reportdir = create_tmpdir_for_user(master, 'report_dir')
+  master_port = 8140
+
+  master_opts = {
```
Comment (whopper) [inline acceptance/tests/reports/failover_master.rb]: This is unused :D

Comment (Magisus) [inline acceptance/tests/reports/failover_master.rb]: And this is what happens when you try to take examples from multiple tests without fully understanding how they work xD

**Comment by whopper** [comment]
@Magisus hm this seems to be hanging on the step where it hits two non-existent masters. Did you see anything like that happen while you were testing?

```
    * master_field should not appear when no master could be conatacted

      f8dqq88udmqord8.delivery.puppetlabs.net (redhat7-64-2) 16:23:45$ puppet agent -t --server_list=badmaster:1,badmaster:2 --report_server=c7ev0dyk145quqo.delivery.puppetlabs.net
        Warning: Could not select a functional puppet master
...
<sits here forever>
```

**Comment by Magisus** [comment]
I noticed that taking longer once or twice, but it always made it through.

**Comment by Magisus** [comment]
Ran this a few more times this morning; it did not hang for me.

**Comment by Magisus** [comment]
Update: when run against most recent puppet-agent builds, the "cannot contact a master" test runs very, very slowly on various Linux distros. The timeouts are much quicker on Windows.

Error output from Windows:
> yd1ca0soynps758.delivery.puppetlabs.net (windows2012r2-64-1) 14:29:16$ cmd.exe /c puppet agent -t --config C:/cygwin64/tmp/puppet_conf_test.r1e8MQ --server_list=badmaster:1,badmaster:2 --report_server=pqspav7i03l9dde.delivery.puppetlabs.net
        Warning: Could not select a functional puppet master
        Info: Retrieving pluginfacts
        Warning: Unable to fetch my node definition, but the agent run will continue:
        Warning: The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Error: /File[C:/ProgramData/PuppetLabs/puppet/cache/facts.d]: Failed to generate additional resources using 'eval_generate': The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Info: Retrieving plugin
        Error: /File[C:/ProgramData/PuppetLabs/puppet/cache/facts.d]: Could not evaluate: Could not retrieve file metadata for puppet:///pluginfacts: The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Error: /File[C:/ProgramData/PuppetLabs/puppet/cache/lib]: Failed to generate additional resources using 'eval_generate': The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Error: /File[C:/ProgramData/PuppetLabs/puppet/cache/lib]: Could not evaluate: Could not retrieve file metadata for puppet:///plugins: The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Error: Could not retrieve catalog from remote server: The requested address is not valid in its context. - connect(2) for "puppet" port 0
        Warning: Not using cache on failed catalog
        Error: Could not retrieve catalog; skipping run
yd1ca0soynps758.delivery.puppetlabs.net (windows2012r2-64-1) executed in 15.87 seconds
      Exited: 1

Error output on Ubuntu:
> ke4hmt1yow20nnr.delivery.puppetlabs.net (ubuntu1604-64-1) 15:30:11$ puppet agent -t --config /tmp/puppet_conf_test.3NLitW --server_list=badmaster:1,badmaster:2 --report_server=j634dps4v23izry.delivery.puppetlabs.net
        Warning: Could not select a functional puppet master
        Warning: Unable to fetch my node definition, but the agent run will continue:
        Warning: execution expired
        Info: Retrieving pluginfacts
        Error: /File[/opt/puppetlabs/puppet/cache/facts.d]: Failed to generate additional resources using 'eval_generate': execution expired
        Error: /File[/opt/puppetlabs/puppet/cache/facts.d]: Could not evaluate: Could not retrieve file metadata for puppet:///pluginfacts: execution expired
        Info: Retrieving plugin
        Error: /File[/opt/puppetlabs/puppet/cache/lib]: Failed to generate additional resources using 'eval_generate': execution expired
        Error: /File[/opt/puppetlabs/puppet/cache/lib]: Could not evaluate: Could not retrieve file metadata for puppet:///plugins: execution expired
        Error: Could not retrieve catalog from remote server: execution expired
        Warning: Not using cache on failed catalog
        Error: Could not retrieve catalog; skipping run
      ke4hmt1yow20nnr.delivery.puppetlabs.net (ubuntu1604-64-1) executed in 783.07 seconds
      Exited: 1

**Comment by joshcooper** [comment]
The variance is likely due to different timeouts when trying to resolve the hostname `badmaster`. You might be able to reduce Puppet's `http_connect_timeout` to shorten the time, but ruby may not be able to interrupt the thread once it calls into the socket APIs.

Alternatively, you could try entering a valid hostname, e.g. localhost, but a nonexistent port (assuming `server_list` will allow you to specify the port). That way you don't incur any DNS resolution delay.

**Comment by Magisus** [comment]
No luck changing the ports and/or hostnames, but setting http_connect_timeout did the trick.

**Comment by branan** [comment]
For two hosts within the same VM cluster (our normal acceptance setup), I'd expect a couple of seconds is more than enough

**Comment by whopper** [comment]
👍  passed in redhat 7, OSX 10.10 and Win 2012R2

**Pull request merged**